Consolidate TextIndicator Layer into WebPageProxy.

megangardner opened the pull request:
#### 37d67d641b0568cf532cf3907a7adda14b4fa794
<pre>
Consolidate TextIndicator Layer into WebPageProxy.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293329">https://bugs.webkit.org/show_bug.cgi?id=293329</a>
<a href="https://rdar.apple.com/151733128">rdar://151733128</a>

Reviewed by Tim Horton.

Now that the textIndicator isn&apos;t in a separate window any
longer, it would be better if the code to add it to a
layer was shared. Thus pushing the code down to
WebPageProxy.

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
(WebKit::WebPageProxy::setTextIndicator):
(WebKit::WebPageProxy::updateTextIndicatorFromFrame):
(WebKit::WebPageProxy::updateTextIndicator):
(WebKit::WebPageProxy::clearTextIndicator):
(WebKit::WebPageProxy::clearTextIndicatorWithAnimation):
(WebKit::WebPageProxy::setTextIndicatorAnimationProgress):
(WebKit::WebPageProxy::teardownTextIndicatorLayer):
(WebKit::WebPageProxy::startTextIndicatoreFadeOut):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy):
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
(WebKit::WebPageProxy::setTextIndicator):
(WebKit::WebPageProxy::updateTextIndicatorFromFrame):
(WebKit::WebPageProxy::updateTextIndicator):
(WebKit::WebPageProxy::clearTextIndicator):
(WebKit::WebPageProxy::setTextIndicatorAnimationProgress):
(WebKit::WebPageProxy::teardownTextIndicatorLayer):
(WebKit::WebPageProxy::startTextIndicatoreFadeOut):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::getTextIndicatorInstallationLayer):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView getTextIndicatorInstallationLayer]):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::getTextIndicatorInstallationLayer):
(WebKit::PageClientImpl::didPerformDictionaryLookup):
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController _animationControllerForText]):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::didPerformDictionaryLookup):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::getTextIndicatorInstallationLayer):

Canonical link: <a href="https://commits.webkit.org/295320@main">https://commits.webkit.org/295320@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b7d2cccab2436e6d953f1a6f77321c8ba69a2ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104743 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14877 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109957 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/55417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79529 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/55417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107749 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/19343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94530 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59836 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54796 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12654 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112357 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31908 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88611 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32272 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90756 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88235 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22485 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33123 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/10900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/27221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31833 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/37186 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31625 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/34966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33184 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12442,6 +12442,11 @@ - (void)_invokeAllActionsToPerformAfterPendingImageAnalysis:(WebKit::ProceedWith
+    return [self layer];
+}
+
 - (void)setUpTextIndicator:(Ref<WebCore::TextIndicator>)textIndicator
```
Why is this (and the code below) still here? I would expect this to be one of the two consolidated pieces.

hortont424 commented on Source/WebKit/UIProcess/mac/PageClientImplMac.h:
```diff
@@ -170,6 +170,7 @@ class PageClientImpl final : public PageClientImplCocoa
     void updateTextIndicator(Ref<WebCore::TextIndicator>) override;
     void clearTextIndicator(WebCore::TextIndicatorDismissalAnimation) override;
     void setTextIndicatorAnimationProgress(float) override;
+    CALayer* getTextIndicatorInstallationLayer() override;
```
no "get" except for out args.

hortont424 replied:
star on the right for ObjC (many places)

aprotyas commented on Source/WebKit/UIProcess/PageClient.h:
```diff
@@ -471,7 +471,8 @@ class PageClient : public CanMakeWeakPtr<PageClient> {
     virtual void clearTextIndicator(WebCore::TextIndicatorDismissalAnimation) = 0;
     virtual void setTextIndicatorAnimationProgress(float) = 0;
-    
+    virtual CALayer* getTextIndicatorInstallationLayer() = 0;
```
Nit: https://www.webkit.org/code-style-guidelines/#pointers-non-cpp
```suggestion
    virtual CALayer *getTextIndicatorInstallationLayer() = 0;
```

hortont424 commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -581,6 +581,11 @@ - (BOOL)_isHostedInAnotherProcess;
     [contentView() setTextIndicatorAnimationProgress:animationProgress];
```
hmm all still here and above too

aprotyas commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
```diff
@@ -161,6 +161,7 @@ class PageClientImpl final : public PageClientImplCocoa
     void updateTextIndicator(Ref<WebCore::TextIndicator>) override;
     void clearTextIndicator(WebCore::TextIndicatorDismissalAnimation) override;
     void setTextIndicatorAnimationProgress(float) override;
+    CALayer* getTextIndicatorInstallationLayer() override;
```
```suggestion
    CALayer *getTextIndicatorInstallationLayer() override;
```

aprotyas commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -581,6 +581,11 @@ - (BOOL)_isHostedInAnotherProcess;
     [contentView() setTextIndicatorAnimationProgress:animationProgress];
 }
 
+CALayer* PageClientImpl::getTextIndicatorInstallationLayer()
```
```suggestion
CALayer *PageClientImpl::getTextIndicatorInstallationLayer()
```

aprotyas replied:
A couple more instances in this PR!

aprotyas commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -295,8 +295,15 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+
+        DictionaryLookup::showPopup(dictionaryPopupInfo, pageClient->viewForPresentingRevealPopover(), [this](TextIndicator& textIndicator) {
+            setTextIndicator(textIndicator.data(), (uint64_t)WebCore::TextIndicatorLifetime::Permanent);
+        }, nullptr, [this]() {
```
nit: We can omit empty lambda parameter lists!
```suggestion
        }, nullptr, [this] {
```

aprotyas commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1503,6 +1504,116 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+    });
+}
+
+void WebPageProxy::setTextIndicator(const WebCore::TextIndicatorData& indicatorData, uint64_t lifetime)
```
Optional: I know you're just moving around code but `lifetime` could just be passed in as a `TextIndicatorLifetime` value right?

rr-codes replied:
strongly agree with Abrar about this

aprotyas commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1503,6 +1504,116 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+void WebPageProxy::updateTextIndicator(const WebCore::TextIndicatorData& indicatorData)
+{
+    if (m_textIndicator && m_textIndicatorLayer)
+        setTextIndicator(indicatorData, 1/*TextIndicatorLifetime::Temporary*/);
```
Since we know we want to pass in `TextIndicatorLifetime::Temporary` (and since we match against this value in `setTextIndicator()`, this may be more appropriate?

```suggestion
        setTextIndicator(indicatorData, std::to_underlying(TextIndicatorLifetime::Temporary));
```

aprotyas replied:
(unless of course you think just changing the argument type to `TextIndicatorLifetime` is more appropriate?)

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1503,6 +1504,116 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+
+void WebPageProxy::clearTextIndicatorWithAnimation(WebCore::TextIndicatorDismissalAnimation animation)
+{
+    RefPtr<WebCore::TextIndicator> textIndicator = WTFMove(m_textIndicator);
```
```suggestion
    RefPtr textIndicator = m_textIndicator;
```

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1503,6 +1504,116 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+{
+    RefPtr<WebCore::TextIndicator> textIndicator = WTFMove(m_textIndicator);
+
+    if ([m_textIndicatorLayer isFadingOut])
+        return;
```
Move early return to the top, so we can avoid the ref if unnecessary.

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1552,6 +1554,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void clearTextIndicator();
+    void clearTextIndicatorWithAnimation(WebCore::TextIndicatorDismissalAnimation);
+    void teardownTextIndicatorLayer();
+    void startTextIndicatoreFadeOut();
```
```suggestion
    void startTextIndicatorFadeOut();
```

hortont424 commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1503,6 +1504,116 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+
+void WebPageProxy::updateTextIndicator(const WebCore::TextIndicatorData& indicatorData)
+{
+    if (m_textIndicator && m_textIndicatorLayer)
```
ideally this would be a early return, IMO

hortont424 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3414,6 +3419,12 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     RunLoop::Timer m_generatePageLoadTimingTimer;
 
+#if PLATFORM(COCOA)
+    RunLoop::Timer m_textIndicatorTimer;
```
timer name should specify what it is about. this is not the text indicator timer, it's the text indicator ... something ... timer. Dismissal? fade? what is it?

megangardner committed:
Consolidate TextIndicator Layer into WebPageProxy.
https://bugs.webkit.org/show_bug.cgi?id=293329
rdar://151733128

Reviewed by Tim Horton.

Now that the textIndicator isn't in a separate window any
longer, it would be better if the code to add it to a
layer was shared. Thus pushing the code down to
WebPageProxy.

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
(WebKit::WebPageProxy::setTextIndicator):
(WebKit::WebPageProxy::updateTextIndicatorFromFrame):
(WebKit::WebPageProxy::updateTextIndicator):
(WebKit::WebPageProxy::clearTextIndicator):
(WebKit::WebPageProxy::clearTextIndicatorWithAnimation):
(WebKit::WebPageProxy::setTextIndicatorAnimationProgress):
(WebKit::WebPageProxy::teardownTextIndicatorLayer):
(WebKit::WebPageProxy::startTextIndicatoreFadeOut):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy):
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
(WebKit::WebPageProxy::setTextIndicator):
(WebKit::WebPageProxy::updateTextIndicatorFromFrame):
(WebKit::WebPageProxy::updateTextIndicator):
(WebKit::WebPageProxy::clearTextIndicator):
(WebKit::WebPageProxy::setTextIndicatorAnimationProgress):
(WebKit::WebPageProxy::teardownTextIndicatorLayer):
(WebKit::WebPageProxy::startTextIndicatoreFadeOut):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::getTextIndicatorInstallationLayer):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView getTextIndicatorInstallationLayer]):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::getTextIndicatorInstallationLayer):
(WebKit::PageClientImpl::didPerformDictionaryLookup):
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController _animationControllerForText]):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::didPerformDictionaryLookup):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::getTextIndicatorInstallationLayer):

Canonical link: https://commits.webkit.org/295320@main
